### PR TITLE
refs [Task][27912] MVP Structure

### PR DESCRIPTION
**Pull request by NguyenTaTrung**
## Related Tickets
- [#Task 27912: MVP structure ](https://edu-redmine.sun-asterisk.vn/issues/27912)

## WHAT
- MVP structure
## HOW
- Create package apply mvp structure
- Create base
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Reviewer1 (TrungNT) | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | Do the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li> 
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseActivity.kt [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+package sun.trainingcourse.anticovid_19.base
 
-import androidx.appcompat.app.AppCompatActivity
 import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
 
-class MainActivity : AppCompatActivity() {
+abstract class BaseActivity : AppCompatActivity() {
+
+    protected abstract val layoutResource: Int
```
Comment (huytq-2422) [inline app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseActivity.kt:10]:
```suggestion
    @get:LayoutRes
    protected abstract val layoutResource: Int
```

File: app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseFragment.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import java.lang.RuntimeException
+
+
+abstract class BaseFragment : Fragment() {
+
+    protected abstract val layoutResource: Int
```
Comment (huytq-2422) [inline app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseFragment.kt:16]:
```suggestion
    @get:LayoutRes
    protected abstract val layoutResource: Int
```

File: app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseFragment.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    protected var fragmentInteractionCallback: FragmentInteractionCallback? = null
```
Comment (huytq-2422) [inline app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseFragment.kt]:
```suggestion
    protected var interactionCallback: FragmentInteractionCallback? = null
```

File: app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseFragment.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        return inflater.inflate(layoutResource, container, false)
+    }
+
+    override fun onAttach(context: Context) {
+        super.onAttach(context)
+        try {
+            fragmentInteractionCallback = context as FragmentInteractionCallback
+        } catch (e: ClassCastException) {
+            throw RuntimeException(context.toString() + " " + FragmentInteractionCallback::class.java.name)
+        }
```
Comment (huytq-2422) [inline app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseFragment.kt]: thay vì try catch, ông có thể check if else 
Nó sẽ nhẹ nhàng hơn rất nhiều

File: app/src/main/res/layout/activity_main.xml [hunk 5]
```diff
@@ -1,18 +1,8 @@
...
-        android:layout_height="wrap_content"
-        android:text="Hello World!"
-        app:layout_constraintBottom_toBottomOf="parent"
-        app:layout_constraintLeft_toLeftOf="parent"
-        app:layout_constraintRight_toRightOf="parent"
-        app:layout_constraintTop_toTopOf="parent" />
-
-</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (huytq-2422) [inline app/src/main/res/layout/activity_main.xml]: conventions 3.1 https://github.com/framgia/coding-standards/blob/master/eng/android/codingstyleguide.md

**Review by huytq-2422 — COMMENTED** [review]
**refs** not **Prefs**

File: app/src/main/res/layout/activity_main.xml (hunk 5 above)
Comment (NguyenTaTrung) [reply app/src/main/res/layout/activity_main.xml]: Tại chưa đến task design nên để trống :D

File: app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseFragment.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    override fun onDetach() {
+        fragmentInteractionCallback = null
+        super.onDetach()
+    }
+
+    interface FragmentInteractionCallback {
+        fun onFragmentInteractionCallBack()
+    }
+
```
Comment (aug16vcc) [inline app/src/main/java/sun/trainingcourse/anticovid_19/base/BaseFragment.kt]: Bỏ dòng trống

File: app/src/main/java/sun/trainingcourse/anticovid_19/base/BasePresenter.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package sun.trainingcourse.anticovid_19.base
+
+interface BasePresenter {
+    fun start() {
+    }
```
Comment (aug16vcc) [inline app/src/main/java/sun/trainingcourse/anticovid_19/base/BasePresenter.kt]:
```suggestion
    fun start()
```

**Review by aug16vcc — APPROVED** [review]

**Pull request merged**